### PR TITLE
[6R2yEhGU] No procedure with name xxxx registered for this database instance

### PR DESCRIPTION
**Pull request by vga91**
WIP: maybe can be related to Trello id: `XWc7tBAb`

- added synchronized to avoid potential conflict
- removed `registerFunction(signature, statement, forceSingle);` and `registerProcedure(signature, statement);` from the system-node merge, to prevent erroneous proc/func deletion before the sys-node is actually created/updated (on lines 224-226, with comment `// de-register removed procs/functions`)
- added tests with multiple procedure/function updates

### CODE REVIEW

File: full/src/test/java/apoc/custom/CypherProceduresClusterTest.java [hunk 1]
```diff
@@ -143,13 +142,93 @@ public void shouldRemoveFunctionOnOtherClusterMembers() throws InterruptedExcept
...
+        // and finally overwrite and re-call it
+        listProcNames.forEach(name -> {
+            String declareProcedure = String.format("CALL apoc.custom.declareProcedure('%s() :: (answer::INT)', $query)", name);
+            String callProcedure = String.format("call custom.%s", name);
+
+            cluster.getSession().writeTransaction(
+                    tx -> tx.run(declareProcedure, Map.of("query", "RETURN 42 AS answer"))
+            );
+
+            // test that it's work for every cluster
```
Comment (conker84) [inline full/src/test/java/apoc/custom/CypherProceduresClusterTest.java:171]:
```suggestion
            // test that it's work for every node
```

File: full/src/test/java/apoc/custom/CypherProceduresClusterTest.java [hunk 2]
```diff
@@ -143,13 +142,93 @@ public void shouldRemoveFunctionOnOtherClusterMembers() throws InterruptedExcept
...
+            cluster.getClusterMembers().forEach(container -> {
+                testCallEventually(container.getSession(), callProcedure, Map.of(),
+                        row -> assertEquals(42L, row.get("answer")),
+                        10L);
+            });
+
+            // overwriting
```
Comment (conker84) [inline full/src/test/java/apoc/custom/CypherProceduresClusterTest.java:178]:
```suggestion
            // overwriting on the leader
```

File: full/src/test/java/apoc/custom/CypherProceduresClusterTest.java [hunk 3]
```diff
@@ -143,13 +142,93 @@ public void shouldRemoveFunctionOnOtherClusterMembers() throws InterruptedExcept
...
+            );
+
+            cluster.getSession().run("call db.clearQueryCaches()");
+
+            // check that it has been updated for every cluster
+            cluster.getClusterMembers().forEach(container -> {
+                testCallEventually(container.getSession(), funQuery, Map.of(),
+                        row -> assertEquals(1L, row.get("row")),
+                        10L);
+            });
```
Comment (conker84) [inline full/src/test/java/apoc/custom/CypherProceduresClusterTest.java:231]: same as above

File: full/src/test/java/apoc/custom/CypherProceduresClusterTest.java [hunk 4]
```diff
@@ -143,13 +142,93 @@ public void shouldRemoveFunctionOnOtherClusterMembers() throws InterruptedExcept
...
+                        row -> assertEquals(42L, row.get("answer")),
+                        10L);
+            });
+
+            // overwriting
+            cluster.getSession().writeTransaction(
+                    tx -> tx.run(declareProcedure, Map.of("query", "RETURN 1 AS answer"))
+            );
+
+            cluster.getSession().run("call db.clearQueryCaches()");
```
[the remark this answers is not in the document]
Comment (vga91) [reply full/src/test/java/apoc/custom/CypherProceduresClusterTest.java]: it's not necessary indeed.
I also removed it from the other cluster tests.

**Comment by vga91** [comment]
Replaced by this pr: https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/3800

**Pull request closed without merging**